qmp/rpc/rpc.go

```diff
@@ -0,0 +1,674 @@
+
+	// Message is an asynchronous notification.
+	Message
+
+	// Stream represents a stream data packet.
+	Stream
+
+	// CallWithFDs is used by a client to indicate the request has
+	// arguments with file descriptors.
+	CallWithFDs
+
+	// ReplyWithFDs is used by a server to indicate the request has
+	// arguments with file descriptors.
+	ReplyWithFDs
+)
+
+// request and response statuses
+const (
+	// StatusOK is always set for method calls or events.
+	// For replies it indicates successful completion of the method.
+	// For streams it indicates confirmation of the end of file on the stream.
+	StatusOK = iota
+
+	// StatusError for replies indicates that the method call failed
+	// and error information is being returned. For streams this indicates
+	// that not all data was sent and the stream has aborted.
+	StatusError
+
+	// StatusContinue is only used for streams.
+	// This indicates that further data packets will be following.
+	StatusContinue
+)
+
+// magic program numbers
+const (
+	programVersion   = 1
+	programRemote    = 0x20008086
+	programQEMU      = 0x20008087
+	programKeepAlive = 0x6b656570
+)
+
+// libvirt procedure identifiers
+const (
+	procConnectOpen        = 1
+	procConnectClose       = 2
+	procDomainLookupByName = 23
+)
+
+// qemu procedure identifiers
+const (
+	qemuDomainMonitor                       = 1
+	qemuConnectDomainMonitorEventRegister   = 4
+	qemuConnectDomainMonitorEventDeregister = 5
+	qemuDomainMonitorEvent                  = 6
+)
+
+const (
+	// packet length, in bytes.
+	packetLengthSize = 4
+
+	// packet header, in bytes.
+	headerSize = 24
+
+	// UUID size, in bytes.
+	uuidSize = 16
+)
+
+// header is a libvirt rpc packet header
+type header struct {
+	// Program identifier
+	Program uint32
+
+	// Program version
+	Version uint32
+
+	// Remote procedure identifier
+	Procedure uint32
+
+	// Call type, e.g., Reply
+	Type uint32
+
+	// Call serial number
+	Serial uint32
+
+	// Request status, e.g., StatusOK
+	Status uint32
+}
+
+// packet represents a RPC request or response.
+type packet struct {
+	// Size of packet, in bytes, including length.
+	// Len + Header + Payload
+	Len    uint32
+	Header header
+}
+
+// libvirt's domain response type
+type domain struct {
+	Name string
+	UUID [uuidSize]byte
+	ID   int
+}
+
+// libvirt domain event
+type event struct {
+	CallbackID   uint32
+	Domain       domain
+	Event        string
+	Seconds      uint64
+	Microseconds uint32
+	Padding      uint8
+	Details      []byte
+}
+
+// internal rpc response
+type response struct {
+	Payload []byte
+	Status  uint32
+}
+
+// RPC implements LibVirt's remote procedure call protocol.
+type RPC struct {
+	// Domain name as seen by libvirt, e.g., stage-lb-1
+	Domain string
+
+	conn net.Conn
+	r    *bufio.Reader
+	w    *bufio.Writer
+
+	// method callbacks
+	cm        sync.Mutex
+	callbacks map[uint32]chan response
+
+	// event listeners
+	em     sync.Mutex
+	events map[uint32]chan *event
+
+	// next request serial number
+	s uint32
```

Comment:
Obviously a `uint32` is quite large, but is there a chance we'd want a `uint64` instead for long-lived request/response cycles?